drivers/counter/counter_mchp_xec.c

```diff
@@ -0,0 +1,413 @@
+						COUNTER_XEC_REG_BASE(dev);
+
+	if (counter->control_reg.enable) {
+		return -EALREADY;
+	}
+
+	counter->control_reg.enable = 1;
+
+	LOG_DBG("Counter started");
+
+	return 0;
+}
+
+static int counter_xec_stop(struct device *dev)
+{
+	volatile struct counter_xec_registers *counter =
+						COUNTER_XEC_REG_BASE(dev);
+
+	if (!counter->control_reg.enable) {
+		return -EALREADY;
+	}
+
+	counter->control_reg.enable        = 0;
+	counter->control_reg.start         = 0;
+	counter->control_reg.halt          = 0;
+	counter->control_reg.reload        = 0;
+	counter->control_reg.auto_restart  = 0;
+
+	counter->int_enable_reg.int_enable = 0;
+	counter->count_reg                 = 0;
+
+	LOG_DBG("Counter stopped");
+
+	return 0;
+}
+
+static u32_t counter_xec_read(struct device *dev)
+{
+	volatile struct counter_xec_registers *counter =
+						COUNTER_XEC_REG_BASE(dev);
+
+	return counter->count_reg;
+}
+
+static int counter_xec_set_alarm(struct device *dev, u8_t chan_id,
+				 const struct counter_alarm_cfg *alarm_cfg)
+{
+	volatile struct counter_xec_registers *counter =
+						COUNTER_XEC_REG_BASE(dev);
+	const struct counter_xec_config *counter_cfg = COUNTER_XEC_CONFIG(dev);
+	struct counter_xec_data *data = COUNTER_XEC_DATA(dev);
+	u32_t ticks;
+
+	if (!counter->control_reg.enable) {
+		return -EIO;
+	}
+
+	if (counter->control_reg.start) {
+		return -EBUSY;
+	}
+
+	if (!alarm_cfg->callback) {
+		return -EINVAL;
+	}
+
+	ticks = alarm_cfg->ticks;
+
+	if (!(alarm_cfg->flags & COUNTER_ALARM_CFG_ABSOLUTE)) {
+		ticks += counter->count_reg;
+	}
+
+	if (counter_cfg->info.max_top_value == UINT16_MAX) {
+		if (ticks > UINT16_MAX) {
+			return -EINVAL;
+		}
+	}
+
+	counter->count_reg = ticks;
+
+	data->alarm_cb = alarm_cfg->callback;
+	data->user_data = alarm_cfg->user_data;
+
+	counter->int_enable_reg.int_enable = 1;
+
+	LOG_DBG("Counter alarm set to %u ticks", ticks);
+
+	counter->control_reg.start = 1;
+
+	return 0;
+}
+
+
+static int counter_xec_cancel_alarm(struct device *dev, u8_t chan_id)
+{
+	volatile struct counter_xec_registers *counter =
+						COUNTER_XEC_REG_BASE(dev);
+	struct counter_xec_data *data = COUNTER_XEC_DATA(dev);
+
+	counter->control_reg.start = 0;
+	counter->int_enable_reg.int_enable = 0;
+
+	data->alarm_cb = NULL;
+	data->user_data = NULL;
+
+	LOG_DBG("Counter alarm canceled");
+
+	return 0;
+}
+
+static u32_t counter_xec_get_pending_int(struct device *dev)
+{
+	volatile struct counter_xec_registers *counter =
+						COUNTER_XEC_REG_BASE(dev);
+
+	return counter->status_reg.event_int;
+}
+
+static u32_t counter_xec_get_top_value(struct device *dev)
+{
+	volatile struct counter_xec_registers *counter =
+						COUNTER_XEC_REG_BASE(dev);
+	const struct counter_xec_config *counter_cfg = COUNTER_XEC_CONFIG(dev);
+
+	if (counter_cfg->info.max_top_value == UINT16_MAX) {
+		return (u32_t)counter->preload16_reg;
+	}
+
+	return counter->preload32_reg;
+}
+
+static int counter_xec_set_top_value(struct device *dev,
```

Comment:
there is a flag `COUNTER_TOP_CFG_DONT_RESET` that should be handled (at list return `-ENOTSUP`)